components/suggestion/search_channel_with_permissions_provider.jsx

```diff
@@ -0,0 +1,237 @@
+            return;
+        }
+
+        const channelsAsync = ChannelActions.autocompleteChannelsForSearch(teamId, channelPrefix)(this.dispatch, this.getState);
+
+        let channelsFromServer = [];
+        try {
+            const {data} = await channelsAsync;
+            channelsFromServer = data;
+        } catch (err) {
+            AppDispatcher.handleServerAction({
+                type: ActionTypes.RECEIVED_ERROR,
+                err,
+            });
+        }
+
+        if (this.shouldCancelDispatch(channelPrefix)) {
+            return;
+        }
+
+        const channels = getChannelsInCurrentTeam(state).concat(channelsFromServer);
+        this.formatChannelsAndDispatch(channelPrefix, resultsCallback, channels);
+    }
+
+    formatChannelsAndDispatch(channelPrefix, resultsCallback, allChannels) {
+        const channels = [];
+
+        const state = this.getState();
+
+        const members = getMyChannelMemberships(state);
+
+        if (this.shouldCancelDispatch(channelPrefix)) {
+            return;
+        }
+
+        const completedChannels = {};
+
+        const channelFilter = this.makeChannelSearchFilter(channelPrefix);
+
+        const config = getConfig(state);
+        const viewArchivedChannels = config.ExperimentalViewArchivedChannels === 'true';
+
+        for (const id of Object.keys(allChannels)) {
+            const channel = allChannels[id];
+
+            if (completedChannels[channel.id]) {
+                continue;
+            }
+
+            if (channelFilter(channel)) {
+                const newChannel = Object.assign({}, channel);
+                const channelIsArchived = channel.delete_at !== 0;
+
+                const wrappedChannel = {channel: newChannel, name: newChannel.name, deactivated: false};
+                if (!viewArchivedChannels && channelIsArchived) {
+                    continue;
+                } else if (!members[channel.id]) {
+                    continue;
+                } else if (channelIsArchived && !members[channel.id]) {
+                    continue;
+                } else if (channel.type === Constants.OPEN_CHANNEL) {
+                    wrappedChannel.type = Constants.OPEN_CHANNEL;
+                } else if (channel.type === Constants.PRIVATE_CHANNEL) {
+                    wrappedChannel.type = Constants.PRIVATE_CHANNEL;
+                } else {
+                    continue;
+                }
+                completedChannels[channel.id] = true;
+                channels.push(wrappedChannel);
+            }
+        }
+
+        const channelNames = channels.
+            sort(channelSearchSorter).
+            map((wrappedChannel) => wrappedChannel.channel.name);
+
+        resultsCallback({
+            matchedPretext: channelPrefix,
+            terms: channelNames,
+            items: channels,
+            component: SearchChannelWithPermissionsSuggestion,
+        });
+    }
+
+    getState = store.getState;
+
+    dispatch = store.dispatch;
```

Comment:
Instead of `this.getState` and `this.dispatch`, just use `store.getState` and `store.dispatch` directly so we're sure it's the latest from store.

Reply:
I have been using this.getState and this.dispatch as a means of mocking the global store in the test the component is referencing the same global store. It is a hacky way to hook into the component to mock it's external dependencies. I wanted to write tests as I was working on the feature. I'll adjust the test to tailor to the mocking of the global store, rather than making the source code tailor to the test.

There are no tests for providers so I was doing what I could to get it under test, but the solution is not maintainable. I have scanned through the tests folder to find if there is a place where the global redux store is mocked, and haven't found it. I'll try and can get my provider test to use a mock global redux store.